Update identifiers to use UUIDs instead of integers

dakoop opened the pull request:
Since integer identifiers make version tree modifications difficult, this branch seeks to replace them with UUIDs.

dakoop committed:
Work in progress on migrating to use UUIDs for identifiers instead of integers.

The goal here is to make things like merging vistrails or dealing with distributed versions of the same vistrail much easier (e.g. for use with crowdLabs).  The main changes so far are a new schema where ids (and foreign ids) are strings that store UUIDs instead of integers and changes to code to support this.

Current known issue is that there is no translation back and forth between the old and new schemas so ***do not use this to create vistrails you wish to open later***.

core/layout/version_tree_layout.py:
  - VistrailsTreeLayoutLW.generateTreeLW: use Vistrail.ROOT_VERSION constant

core/modules/module_descriptor.py:
  - ModuleDescriptor.add_input_port: include self when computing sort_key
  - ModuleDescriptor.add_output_port: include self when computing sort_key

core/modules/module_registry.py:
  - ModuleRegistry.add_port: compute sort key based on the length of the port_specs_list since sorting the id does not work any longer

core/vistrail/controller.py:
  - VistrailController._set_current_version: raise Exception is 0 is passed
  - VistrailController.flush_pipeline_cache: use Vistrail.ROOT_VERSION
  - VistrailController.set_version: use Vistrail.ROOT_VERSION
  - VistrailController.create_port_spec_static: compute sort key based on length of the port_specs_list since sorting by id does not work any longer
  - VistrailController.recompute_terse_graph: use Vistrail.ROOT_VERSION
  - VistrailController.do_version_switch: use Vistrail.ROOT_VERSION, don't convert tag to long

core/vistrail/pipeline.py:
  - Pipeline.get_tmp_id: do not need negative sign here (just a UUID)

core/vistrail/vistrail.py:
  - Vistrail.getPipeline: can't use dispatcher here since UUIDs are also strings, just check if the tag exists
  - Vistrail.getVersionGraph: **removed**
  - Vistrail.pruneVersion: **moved** to gui.vistrail_controller
  - Vistrail.getPipelineDispatcher: **removed**
  - ExplicitExpandedVersionTree: use Vistrail.ROOT_VERSION
  - TestVistrail.test_version_graph: **removed**

db:
  - added version v2_0_0 which uses UUIDs for ids, changed various pointers to reflect this

db/services/action_chain.py:
  - getActionChain: use != not <
  - get_reduced_operations: new method
  - get_operation_diff: new method
  - getCurrentOperations: use get_reduced_operations

db/services/vistrail.py:
  - materializeWorkflow: use DBVistrail.ROOT_VERSION
  - performAdds: ignore "None-ops"
  - invertOperations: use reversed instead of in-place reverse, ignore "None-ops"
  - normalOperations: ignore "None-ops"
  - getPathAsAction: use updated utility routines, keep adds, deletes separate
  - getSharedRoot: changed to make more sense and support new id scheme
  - getOperationDiff: use in instead of has_key

gui/collection/vis_log.py:
  - QExecutionListWidget.add_workflow_exec: don't convert parent_version to int

gui/pipeline_view.py:
  - QAbstractGraphicsPortItem.__init__: don't set a default z-value
  - QAbstractGraphicsPortItem.mouseMoveEvent: calculate the z-value using the module's z-values
  - QGraphicsConfigureItem.__init__: don't set a default z-value
  - QGraphicsConnectionItem.__init__: set z-value using module's z-values
  - QGraphicsModuleItem.setupModule: set z-value to 0.5, need to fix

gui/query_view.py:
  - QQueryView.build_widget: use new signal for versionSelected

gui/version_view.py:
  - QGraphicsVersionItem.show_raw_pipeline: new signal for versionSelected
  - QGraphicsVersionItem.construct_from_root: new signal for versionSelected
  - QVersionTreeScene.adjust_version_colors: use Vistrail.ROOT_VERSION
  - QVersionTreeScene.setupScene: use Vistrail.ROOT_VERSION
  - QVersionTreeScene.selectedChanged: use new signal for versionSelected
  - QVersionTreeScene.double_click: use new signal for versionSelected

gui/vistrail_controller.py:
  - VistrailController._change_version_short_hop: use Vistrail.ROOT_VERSION
  - VistrailController.show_parent_version: use Vistrail.ROOT_VERSION
  - VistrailController.prune_version: **moved** from core.vistrail.vistrail.Vistrail, use Vistrail.ROOT_VERSION
  - VistrailController.prune_versions: use Vistrail.ROOT_VERSION, prune_version call
  - VistrailController.hide_versions_below: use _current_full_graph instead of getVersionGraph
  - VistrailController.show_all_versions: use _current_full_graph instead of getVersionGraph
  - VistrailController.expand_versions: use _current_full_graph instead of getVersionGraph, != instead of >
  - VistrailController.collapse_versions: use _current_full_graph instead of getVersionGraph
  - VistrailController._expand_or_collapse_all_versions_below: use _current_full_graph instead of getVersionGraph
  - VistrailController.get_pipeline_name: use Vistrail.ROOT_VERSION

gui/vistrail_view.py:
  - QVistrailView.create_version_view: use new versionSelected signal

dakoop committed:
<bugfix>Fix typo for port spec sort key</bugfix>

core/vistrail/controller.py:
  - VistrailController.create_port_spec_static: port_spec_key -> port_sort_key

dakoop committed:
Merge branch 'master' into use-uuid

dakoop committed:
Merge branch 'master' into use-uuid

dakoop committed:
Update schema to match entities in v1.0.3 schema.

Changes ids on mashups to use uuids as well.

dakoop committed:
<bugfix>Fix QChar issue with parameter exploration</bugfix>

Use unichr instead of QChar.  Note, however, that this looks bizarre on Mac as it shows a large triangle and a small triangle.

vistrails/gui/modules/paramexplore.py:
  - QParameterEditorSelector.__init__: use unichar instead of QChar

dakoop committed:
<feature>Display version id in version properties</feature>

Seeing how much depends on version ids, I think it is nice to actually be able to see the version ids.

versions/gui/version_prop.py:
  - QVersionProp.__init__: add id labels
  - QVersionProp.updateVersion: update id field

dakoop committed:
Update id display for UUIDs.

vistrails/gui/common_widgets.py:
  - QElideLabel: new class

vistrails/gui/version_prop.py:
  - QVersionProp.__init__: use QElideLabel for id
  - QVersionProp.updateVersion: no need to cast to unicode

dakoop committed:
Merge branch 'master' into use-uuid

Conflicts:
	vistrails/gui/collection/vis_log.py
	vistrails/gui/version_prop.py
	vistrails/gui/vistrail_controller.py

dakoop committed:
WIP: Translate older vistrails to use uuids

This commit begins the process of translating older vistrails (schema 1.0.3) to use uuids.  Known issues:

(1) The link between workflow execs and workflows is not working.  We have support for translating individual entities like vistrails, workflows, logs, but not entire bundles.  This will need to be added.
(2) Some entities that were changed to uuids (session) are not translated because they are not "ids" so the copy won't reassign ids.
(3) The translation is a bit hacky as it uses do_copy from 2.0.0 objects but passing 1.0.3 objects; this requires hacking __class__ which isn't great.

vistrails/core/collection/vistrail.py:
  - VistrailEntity.add_workflow_entity: use ROOT_VERSION, not 0
  - VistrailEntity.add_wf_exec_entity: use ROOT_VERSION, not 0
  - VistrailEntity.reload: if get_persisted_log fails, need to already have log set to None for the guard below to catch

vistrails/core/vistrail/vistrail.py:
  - Vistrail.get_pipeline_name: raise exception is version doesn't exist

vistrails/db:
  - change session in wf_exec to be a uuid (matching vistrail)

vistrails/db/versions/__init__.py:
  - translate_object: add 1.0.3 -> 2.0.0 and 2.0.0 -> 1.0.3 translations

vistrails/db/versions/v1_0_3/translate/v2_0_0.py:
  - initial idea to translate back from uuids

vistrails/db/versions/v2_0_0/translate/v1_0_3.py:
  - translate from numeric ids to uuids

vistrails/gui/collection/vis_log.py:
  - QLogView.set_controller: allow db_name to be "Unknown" so we can open old vistrails

dakoop committed:
Merge branch 'master' into use-uuid

Conflicts:
	vistrails/core/modules/module_descriptor.py
	vistrails/db/domain/__init__.py
	vistrails/db/persistence/__init__.py
	vistrails/db/specs/all.xml

dakoop committed:
Fixes to match updates to schema.

dakoop committed:
Merge branch 'master' into use-uuid

dakoop committed:
Fix vistrails.core.api unit tests.

dakoop committed:
Fix unit tests for ActionAnnotations.

dakoop committed:
Merge branch 'master' into use-uuid

Conflicts:
	vistrails/core/api.py
	vistrails/core/collection/vistrail.py
	vistrails/core/modules/module_registry.py
	vistrails/core/vistrail/controller.py
	vistrails/db/domain/__init__.py
	vistrails/db/persistence/__init__.py
	vistrails/db/services/action_chain.py
	vistrails/db/versions/__init__.py
	vistrails/gui/common_widgets.py
	vistrails/gui/module_info.py
	vistrails/gui/pipeline_view.py
	vistrails/gui/version_prop.py
	vistrails/gui/version_view.py
	vistrails/gui/vistrail_controller.py

dakoop committed:
WIP: Updates to get basic GUI functionality working

dakoop commented:
There are a lot of tests that break with this right now. Much more work to be done.

dakoop committed:
Add test support for UIUDs

dakoop committed:
Fix a few issues with int casts and getSharedRoot

dakoop committed:
Move reverse 2.0.0 translation from 1.0.3 to 1.0.4

dakoop committed:
Fix layout bug (0 vs. ROOT_VERSION)

dakoop committed:
Remap upgrade annotations

dakoop committed:
Fix version strings

dakoop committed:
Update logic for version tree operations

dakoop committed:
Add action comparison operators (compare by date)

remram44 commented:
Things to watch out for: reading IDs from annotations (for instance, upgraded versions) probably currently always involve a str->int cast.

Also, some methods (at least API) accept either tag name or version number, and these are usually discriminated by type.

dakoop commented:
Yes, I was also thinking about modifying the db objects/calls to check types to catch these cases, but it is true that some of the logic is at the core level. I want to roll in bundle/sqlalchemy support at the same time since we need to have a "bundle upgrade" for this type of translation (e.g. ids in the log need to be remapped to match the ids in the vistrail).

dakoop committed:
Merge branch 'v2.2' into use-uuid

dakoop committed:
Update logic for uuids

dakoop committed:
More zero to Vistrail.ROOT_VERSION changes

dakoop committed:
Refactor Vistrail.get_version_number to Vistrail.get_version_id

dakoop committed:
Remove/replace int/long casts of versions

dakoop committed:
Fix for UUID

dakoop committed:
Add versions.get_current_version as method, refactor io

dakoop committed:
Update db id remapping logic

dakoop committed:
Start script to convert vts to uuids and recover mapping

dakoop committed:
Fix upgradeworkflow tests for uuids

dakoop committed:
Fix change_selected_version(0) calls

Use Vistrail.ROOT_VERSION instead

dakoop committed:
Allow -1 as version_id in get_version_id

dakoop committed:
Update controlParameter id to uuid

dakoop committed:
Fix do_copy with new_ids issue

Potentially a major issue where copied ids were not assigned correctly.

dakoop committed:
Update uuid translation logic

dakoop committed:
Fix issues related to current_version=-1

dakoop committed:
Remove sorting-by-id logic in diff, use OrderedDict

dakoop committed:
Change more action chains/diff logic to use OrderedDict

dakoop committed:
Add uuid version of dummy.xml

dakoop committed:
Make sure field order is maintained from spec

This is important because if we copy an entity, the code needs to be
aware of any changes to ids that may be in the same object
(e.g. DBWorfklowExec.item_execs references ids in
DBWorkflowExec.machines via the machine_id foreign key, but before the
change, machines were added after item_execs). This is only an issue
when choices need to be after properties.

dakoop committed:
Add updated v2.0.0 domain classes

dakoop committed:
Fix unit test

dakoop committed:
Add support for converting log to uuids

Carry ids from vistrail entities over to log

dakoop committed:
Use autopep8 for db generate code

dakoop committed:
Change format flag for uuid

dakoop committed:
Fix workflow exec foreign key refs

dakoop committed:
Fix issue where JobMonitor expects KeyError from get_pipeline_name

dakoop committed:
Add translation for mashups, update convert uuid script

dakoop committed:
Rename registry's idScope -> id_scope

dakoop committed:
Fix mashups to better adhere to db standards

dakoop committed:
Update mashup foreign keys into vistrails

Needed for accurate translation

dakoop committed:
More mashups updates to better work with db structure

dakoop committed:
Update jobMonitor test vistrail (with log and mashups)

dakoop committed:
Fix create_group test

dakoop committed:
Fix diff logic for uuid

dakoop committed:
Add db mashup root class for 1.0.3, update to fix mashup test

dakoop committed:
Remove version_tag and Locator._vtag

Use version_node for all version naming

dakoop committed:
Fix tests with minor issues

dakoop committed:
Update upgrade tests in gui.vistrail_controller

dakoop committed:
Fix typo

dakoop committed:
Check ROOT_VERSION not 0

dakoop committed:
Update graph upgrade tests in controller

dakoop committed:
Allow update_version to skip updating the id scope

dakoop committed:
Migrate translation tests from sql-alchemy branch

Also add a lot of infrastructure for doing the correct translations
and allowing remaps to be maintained both across bundles and used for
tests

dakoop committed:
Add missing test_basics file

dakoop committed:
Make uuids work	with SQL backend

 - Remove tags field
 - Fix issue with vistrail variables: uuid must be the key and
   id scope must know about it
 - Fix a few inverse refs that were typed as integers
 - Update the 2.0.0 persistence files
 - Update MySQL tests to run for 1.0.4 and 2.0.0

dakoop committed:
Update 2.0.0 copyright and xml processing code

dakoop committed:
Add missing version tests

dakoop committed:
Add code to help identify id overlaps

dakoop committed:
Add bundle support and translation for versions back to 1.0.1

dakoop committed:
Fix machine references in 1.0.3 <-> 1.0.4

Machine moved from Log to individual WorkflowExecs

dakoop committed:
Update translation tests to test more

Uses metaclass to create and update tests

dakoop committed:
Add test to demonstrate critical issue with create_copy_op_chain

dakoop committed:
Fix #1191

Make sure children of any object that were not orphaned are assigned
new ids

dakoop committed:
Add pos field for operation in 2.0.0 schema

Indicates position. Required because we cannot sort by id as uuids
will not be sequential and db may not maintain order.

dakoop committed:
Fix issue related to #1191 by reassigning PortSpecItem ids

Note that because PortSpecItem ids are not referenced (except when
serializing to the database), just reassign them to be unique before
they are translated to uuids

dakoop committed:
Fix id scope issue related to #1191

Must update the id scope with DBPortSpecItem ids

dakoop committed:
Change how bundle object is determined

Use attribute check

dakoop committed:
Fix issue where log was translated even if file didn't exist

dakoop committed:
Clean up io tests

dakoop committed:
Create common translate methods that can be specialized

Cases where an object does not modify fields now do not require the
boilerplate translate code

dakoop committed:
Fix string format bug

dakoop committed:
Fix bug introduced in PR #1169

Need to make sure the remapped pipeline is processed with the remapped
errors

dakoop committed:
Fix issue with log translation

dakoop committed:
Add missing updates to common translate code

dakoop committed:
Make individual VTL tests

Use a metaclass to create individual test functions

dakoop committed:
Remove old API tests

dakoop committed:
Use common save bundle

dakoop committed:
Add abstraction translation

Internal versions in the vistrail need to be updated so we translate
abstractions first to know what this remap is

dakoop committed:
Move recursive remapping to class ExternalData

Add support for subworkflow translations

dakoop committed:
Fix key issue in subworkflow id remap

dakoop committed:
Use common SaveBundle

dakoop committed:
Also use db_version to get object's version

dakoop committed:
Fix pipeline tests

dakoop committed:
Fix 1.0.3 -> 1.0.2 translation tests

dakoop committed:
Fix vistrails_module tests to use tags instead of ids

dakoop committed:
Update 1.0.2 -> 1.0.3 translation tests

dakoop committed:
Update bundle test to include mashups, subworkflows

dakoop committed:
Merge branch 'v2.2' into use-uuid

dakoop committed:
Update abstraction/subworkflow creation to use uuids

dakoop committed:
Merge branch 'master' into use-uuid

dakoop committed:
Fix typo

dakoop committed:
Fix core -> GUI requirements mis-merge

dakoop committed:
Update v2.0.0 schema with the 1.0.5 union change

dakoop committed:
Fix core -> gui requirements in dialogs pkg

dakoop committed:
Fix port spec union attribute for sql (reserved word)

dakoop committed:
Ignore port spec union test from 1.0.4 -> 2.0.0

dakoop committed:
Fix schema issue with incorrect entity_id type

dakoop committed:
Regenerate sql code for entity_id change

dakoop committed:
Update global specs with ps_union change

dakoop committed:
Fix issue with port spec order

Cannot rely on port spec uuids to order specs
